Add factory reset after 5 reboots example to cookbook (#3433)
OttoWinter/esphomedocs: cookbook/lambda_magic.rst

        name: "Number from text"

Factory reset after 5 quick reboots
-----------------------------------

> One may want to restore factory settings (eg. Wi-Fi credentials set at runtime, or clear restore states) without having to
> disassemble or dismount the devices from their deployed location, whilst there's no network access either. The example below
> shows how to achieve that using lambdas in a script by triggerig the factory reset switch after the system rebooted 5 times
> with 10 seconds timeframes.

Reviewer: ```suggestion
One may want to restore factory settings (like Wi-Fi credentials set at runtime, or clear restore states) without having to
disassemble or dismount the devices from their deployed location, whilst there's no network access either. The example below
shows how to achieve that using lambdas in a script by triggering the factory reset switch after the system rebooted 5 times
with 10-second timeframes.
```

.. code-block:: yaml

    # Example config.yaml
    esphome:
      name: "esphome_ld2410"
      on_boot:
        priority: 600.0
        then:
          - script.execute: fast_boot_factory_reset_script
    esp32:
      board: esp32-c3-devkitm-1

    substitutions:
      factory_reset_boot_count_trigger: 5 

    globals:
      - id: fast_boot
        type: int
        restore_value: yes
        initial_value: '0'
    
    script:
      - id: fast_boot_factory_reset_script
        then:
          - if:
              condition:
                lambda: return ( id(fast_boot) >= ${factory_reset_boot_count_trigger});
              then:
                - lambda: |-
                    ESP_LOGD("Fast Boot Factory Reset", "Performing factotry reset");
                    id(fast_boot) = 0;
                    fast_boot->loop();
                    global_preferences->sync();

>                 - switch.turn_on: factory_reset_switch

Reviewer 2: ```suggestion
                - button.press: factory_reset_button
```

          - lambda: |-
              if(id(fast_boot) > 0)
                ESP_LOGD("Fast Boot Factory Reset", "Quick reboot %d/%d, do it %d more times to factory reset", id(fast_boot), ${factory_reset_boot_count_trigger}, ${factory_reset_boot_count_trigger} - id(fast_boot));
              id(fast_boot) += 1;
              fast_boot->loop();
              global_preferences->sync();
          - delay: 10s
          - lambda: |-
              id(fast_boot) = 0;
              fast_boot->loop();
              global_preferences->sync();
    
    wifi:
      id: wifi_component
      ap:
        ap_timeout: 0s
      reboot_timeout: 0s
    
    captive_portal:
    

>     switch:

Reviewer 2: ```suggestion
    button:
```

      - platform: factory_reset

>         id: factory_reset_switch

Reviewer 2: ```suggestion
        id: factory_reset_button
```

        name: "Number from text"

Factory reset after 5 quick reboots
-----------------------------------

One may want to restore factory settings (eg. Wi-Fi credentials set at runtime, or clear restore states) without having to
disassemble or dismount the devices from their deployed location, whilst there's no network access either. The example below
shows how to achieve that using lambdas in a script by triggerig the factory reset switch after the system rebooted 5 times
with 10 seconds timeframes.

.. code-block:: yaml

    # Example config.yaml
    esphome:
      name: "esphome_ld2410"
      on_boot:
        priority: 600.0
        then:
          - script.execute: fast_boot_factory_reset_script
    esp32:
      board: esp32-c3-devkitm-1
    
    globals:
      - id: fast_boot
        type: int
        restore_value: yes
        initial_value: '0'
    
      - id: factory_reset_boot_count_trigger
        type: int

>         initial_value: '5'

Reviewer 2: This doesn't really make sense as a global, it could just be a substitution

Author: Done
